safe_app/src/tests/mod.rs

```diff
@@ -32,14 +31,114 @@ use safe_core::{Client, CoreError};
 use safe_nd::{
     ADataAddress, ADataOwner, AppPermissions, AppendOnlyData, Coins, Error as SndError,
     PubImmutableData, PubSeqAppendOnlyData, PubUnseqAppendOnlyData, UnpubUnseqAppendOnlyData,
-    XorName,
+    XorName, GET_BALANCE, PERFORM_MUTATIONS, TRANSFER_COINS,
 };
 #[cfg(feature = "mock-network")]
 use safe_nd::{RequestType, Response};
 use std::collections::HashMap;
 use std::rc::Rc;
 use unwrap::unwrap;
 
+// App Tokens test:
+// 1. Authenticate an app A.
+// 2. Send request as app A assert that tokens are valid.
+// 3. Re-authenticate app A again with new permissions so that token gets updated.
+// 4. Assert if the tokens are updated by sending the older token - should fail.
+// 5. Assert again with the new token - should pass.
+#[test]
+pub fn app_tokens_test() {
+    let auth = authenticator::create_account_and_login();
+    let app_info = authenticator::rand_app();
+
+    // Authenticate App A
+    let auth_response = unwrap!(authenticator::register_app(
+        &auth,
+        &AuthReq {
+            app: app_info.clone(),
+            app_container: false,
+            app_permissions: Default::default(),
+            containers: HashMap::new(),
+        },
+    ));
+
+    let the_token = auth_response.clone().token;
+    // Verify the token
+    unwrap!(auth_run(&auth, move |client| {
+        // Verify if the signature is signed by the client and is verifiable by client's public ID
+        assert!(the_token
+            .is_valid_for_public_key(&client.public_id().public_key())
+            .is_ok());
+
+        // Verify the contents in the Caveat
+        fn caveat_is_false_checker(contents: String) -> bool {
+            contents.as_str() == "false"
+        }
+
+        assert!(the_token
+            .verify_caveat(GET_BALANCE, caveat_is_false_checker)
+            .is_ok());
+        assert!(the_token
+            .verify_caveat(PERFORM_MUTATIONS, caveat_is_false_checker)
+            .is_ok());
+
+        assert!(the_token
+            .verify_caveat(TRANSFER_COINS, caveat_is_false_checker)
+            .is_ok());
+        Ok(())
+    }));
+
+    let app = unwrap!(App::registered(app_info.id.clone(), auth_response, || ()));
+
+    let app_permissions = AppPermissions {
+        transfer_coins: true,
+        perform_mutations: false,
+        get_balance: true,
+    };
+
+    // Re-authenticate the app with new permissions
+    let auth_response_new = unwrap!(authenticator::register_app(
+        &auth,
+        &AuthReq {
+            app: app_info.clone(),
+            app_container: false,
+            app_permissions,
+            containers: HashMap::new(),
+        },
+    ));
+
+    // Generates new App instance for the the re-authenticated app
+    let app_new = unwrap!(App::registered(app_info.id, auth_response_new, || ()));
+
+    // Send requests to check app permissions from old app instance - should fail
+    unwrap!(run(&app, move |client, _| {
+        client
+            .get_balance(None)
```

Comment:
The app never had these perms, so should always fail.

Here we want to:

- Auth an app w/ perms (get balance eg).
- check that goes 
- auth with _new_ perms.
- Check that first app token is _no longer valid_ (as a new token has been reissued)
- check our new token is valid.